homeassistant/components/usb/utils.py

```diff
@@ -0,0 +1,34 @@
+"""The USB Discovery integration."""
+from __future__ import annotations
+
+from typing import Optional, Tuple
+
+from serial.tools.list_ports_common import ListPortInfo
+
+from .models import USBDevice
+
+USBDeviceTupleType = Tuple[str, str, str, Optional[str], Optional[str], Optional[str]]
+
+
+def usb_device_from_port(port: ListPortInfo) -> USBDevice:
+    """Convert serial ListPortInfo to USBDevice."""
+    return {
+        "device": port.device,
+        "vid": f"{hex(port.vid)[2:]:0>4}".upper(),
+        "pid": f"{hex(port.pid)[2:]:0>4}".upper(),
+        "serial_number": port.serial_number,
+        "manufacturer": port.manufacturer,
+        "description": port.description,
+    }
+
+
+def usb_device_tuple(usb_device: USBDevice) -> USBDeviceTupleType:
+    """Generate a unique tuple for a usb device."""
+    return (
```

Comment:
If we make `USBDevice` a dataclass we can use `dataclasses.astuple` to get a tuple.

https://docs.python.org/3/library/dataclasses.html#dataclasses.astuple

Reply:
Good idea. Thats a much better implementation 👍 

Will clean this up after coffee 😄

Reply:
Done